### ISSUE TITLE
JIT loses access to local variables

### ISSUE THREAD

**Issue by hzyjerry** [google/jax#1426]
Hi, thanks for your effort!

I'm currently implementing some optimal control script with jax, and I find the following function very helpful
```python
def test_retain():
    a = 10
    def loss_sum(val):
        loss = np.sum(val[:2]**2 + val[2:]**3) * a
        return loss
    grad = jax.grad(loss_sum)
    gs = grad(np.array([2, 3, 4, 5], dtype=float))
    assert np.allclose(gs, np.array([  40.,  60.,  480., 750.]))
    a = 20
    gs = grad(np.array([2, 3, 4, 5], dtype=float))
    assert np.allclose(gs, np.array([  80.,  120.,  960., 1500.]))
```
In a sense, the loss function retains access to variable `a` outside, which allows me to change `a` without recomputing the same graph every time

However it fails with jit compilation
```python
def test_retain():
    a = 10
    def loss_sum(val):
        loss = np.sum(val[:2]**2 + val[2:]**3) * a
        return loss
    grad = jax.jit(jax.grad(loss_sum))     # added JIT
    gs = grad(np.array([2, 3, 4, 5], dtype=float))
    assert np.allclose(gs, np.array([  40.,  60.,  480., 750.]))
    a = 20
    gs = grad(np.array([2, 3, 4, 5], dtype=float))
    assert np.allclose(gs, np.array([  80.,  120.,  960., 1500.])) # fails
```

I wonder if this is a fundamental limit of JIT? Do you have suggestions to use JIT for speeding up the code in this case?

**Comment by hzyjerry** [comment]
Oh I realized that I should call `jax.jit` the second time I run the gradient operation. Closing the issue for now.
